File: 1396-design-underground-system/1396-design-underground-system.cpp
/**
* Leetcode Problem - https://leetcode.com/problems/design-underground-system/
* Author - Shubham Nagaria
* Date - 24th April 2022
Design Underground System
**/
class UndergroundSystem {
public: 
    unordered_map<string,vector<double>>avgSumcountSourceDestination;
    unordered_map<int,pair<string,int>>checkin;
    UndergroundSystem() {
        avgSumcountSourceDestination.clear();
        checkin.clear();
    }
    
    void checkIn(int id, string stationName, int t) {
        checkin[id]={stationName,t};
    }
    
    void checkOut(int id, string stationName, int t) {
            auto idxCheckin = checkin[id];
            string key = idxCheckin.first +"_"+stationName;
            double time = t - idxCheckin.second;
            if(avgSumcountSourceDestination.find(key)!= avgSumcountSourceDestination.end()){
                avgSumcountSourceDestination[key][1] += time;
                avgSumcountSourceDestination[key][2]++;
                avgSumcountSourceDestination[key][0]= avgSumcountSourceDestination[key][1]/avgSumcountSourceDestination[key][2];
            }
            else{
                avgSumcountSourceDestination[key]={time,time,1};
            }
    }
    
    double getAverageTime(string startStation, string endStation) {
        string key = startStation +"_" + endStation;
        return avgSumcountSourceDestination[key][0];
    }
};

/**
 * Your UndergroundSystem object will be instantiated and called as such:
 * UndergroundSystem* obj = new UndergroundSystem();
 * obj->checkIn(id,stationName,t);
 * obj->checkOut(id,stationName,t);
 * double param_3 = obj->getAverageTime(startStation,endStation);
 */

File: 1396-design-underground-system/1396-design-underground-system.py
class UndergroundSystem:
    def __init__(self):
        self.avgSumCountSrcDes={}
        self.checkin={}
    def checkIn(self, id: int, stationName: str, t: int) -> None:
        self.checkin[id]=list()
        self.checkin[id].append(stationName)
        self.checkin[id].append(t)

    def checkOut(self, id: int, stationName: str, t: int) -> None:
        p = self.checkin[id]
        key = p[0]+"_"+stationName
        if key in self.avgSumCountSrcDes:
            self.avgSumCountSrcDes[key][1] += t-p[1]
            self.avgSumCountSrcDes[key][2] +=1
            self.avgSumCountSrcDes[key][0]= self.avgSumCountSrcDes[key][1]/self.avgSumCountSrcDes[key][2]
        else:
            self.avgSumCountSrcDes[key]=[t-p[1],t-p[1],1]
    
    def getAverageTime(self, startStation: str, endStation: str) -> float:
        key = startStation+"_"+endStation
        return self.avgSumCountSrcDes[key][0]


# Your UndergroundSystem object will be instantiated and called as such:
# obj = UndergroundSystem()
# obj.checkIn(id,stationName,t)
# obj.checkOut(id,stationName,t)
# param_3 = obj.getAverageTime(startStation,endStation)